Attempt to use Github Actions for CI

tmccombs opened the pull request.

tmccombs committed:
Attempt to use Github Actions for CI

SrodriguezO commented on publish.sh:
```diff
@@ -2,10 +2,13 @@
 fi
-echo "Running: sbt ++$TRAVIS_SCALA_VERSION \"$command\""
-exec sbt ++$TRAVIS_SCALA_VERSION "$command"
+export SBT_OPTS="-Dbuild.version=$version"
```
~~why do we need this now?~~ nvm, it comes from the `before_script` of the travis config

SrodriguezO commented on build.sbt:
```diff
@@ -40,7 +40,6 @@ lazy val specs2Dependency = Seq(
 val catsVersion = "2.+"
 lazy val catsDependency = Seq(
   "org.typelevel" %% "cats-macros" % catsVersion,
-  "org.typelevel" %% "cats-kernel" % catsVersion,
```
what's this change for?

tmccombs replied:
I was getting a warning about conflicting versions or something like that, and I realized that cats-kernel is pulled in by cats-core, so I didn't need to explicitly depend on it.

SrodriguezO commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,78 @@
+          SCALA_VERSION: ${{ matrix.scala_version }}
+          PGP_SECRET: ${{ secrets.PGP_SECRET }}
+  release-upload:
+    if: github.event_name == "release"
```
iiuc, this job automatically creates a release for this tag. But this only runs on the `release - created` event; did you mean to have it run on tag push?

SrodriguezO replied:
Oh... this job edits the existing release. I see. Should we have the job run on tag push and create the release too? It seems that was what Travis was doing.

SrodriguezO commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,78 @@
+          path: '*/target/'
+          retention-days: 1
+  publish:
+    if: github.event_name == "push" || github.event_name == "release"
```
similarly, should this be on tag/master push?

tmccombs replied:
this workflow is only triggered on pushes to master and when releases are created. And pull requests. I suppose, I could change this to `github.event_name != 'pull_request'`

SrodriguezO replied:
Right, should the workflow trigger on release-create or on tag push?

SrodriguezO commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,78 @@
+    steps:
+      - uses: actions/checkout@v2
+      - uses: actions/cache@v2
+        with: &cache_options
```
I have not seen this syntax before; it seems to let you re-use options? I can't find documentation on these `&` and `*` symbols for GHA; do you have any links?

SrodriguezO replied:
Ah.. it seems these aren't currently supported yet. It's fialing to start the CI: https://github.com/lucidsoftware/xtract/actions/runs/435208437

`The workflow is not valid. .github/workflows/ci.yml: Anchors are not currently supported. Remove the anchor 'scala_versions'`

SrodriguezO replied:
Ah.. it seems these aren't currently supported yet. It's failing to start the CI: https://github.com/lucidsoftware/xtract/actions/runs/435208437

`The workflow is not valid. .github/workflows/ci.yml: Anchors are not currently supported. Remove the anchor 'scala_versions'`

tmccombs replied:
:( that's unfortunate. I was trying to avoid having to repeat the same config in multiple places. My biggest gripe with GHA is it can sometimes be very verbose, and there isn't a good way to avoid duplication.

SrodriguezO commented on .travis.yml:
```diff
@@ -1,50 +0,0 @@
-deploy:
-  - provider: releases
-    api_key:
-      secure: NaItKK8ZQUbMJUDqtSAgfE0idFF2dDb27yGvWZYT6K9E2DWaSjY4GCQM6uhg2TkaiyMMcsy+19nls8iHm/tfgrU6yLRhc3TpEDL6MTi+Db1LsXQbu6gpOhBEl5K0352lIi0RqEeVWpDiPT+kErg2U/Njq0TSkuAnGAUNYMgeRuc0cGgR5np7dYOSJpW9TO4lV9NDmkALl8EnVx94TlXlo94rsht1Io9Gse7gnRajXdyyk+iBAckATodOeBN42Zg/6ymUwhzeaEn6imVFiUSW/4WYYZyoNEvrevBw6+tqdg9MM10CkRcYe1cnos0rsMm1/05QE3IszhtOZ1lSay8P0xUS615Kmgo751Kf6s5oBeylx5MIQAroS4VvHo4XrojiA ...
```
where was this API key coming from? Since we don't need it anymore, can we delete it?

tmccombs replied:
It's a github api key generated by travis. There isn't really anything to delete besides revoking access to travis once everything is moved to github actions.

SrodriguezO commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,78 @@
+          set -x
+          tag="${GITHUB_REF#refs/tags/}"
+          args=$(printf -- "-a %q " {xtract-core,testing,macros}/target/*/*.jar)
+          hub release edit $args -m "" $tag
```
we weren't publishing `macros/target/*/*.jar` before; double checking this change intentional

tmccombs replied:
yep. It should have been published, but wasn't.

SrodriguezO commented on .travis.yml:
```diff
@@ -1,50 +0,0 @@
-install: sbt ++$TRAVIS_SCALA_VERSION update
-before_cache:
-  #Avoid unncessary cache updates
-  - find $HOME/.ivy2 -name "ivydata-*.properties" -print -delete
```
is there a way to do this with the new cache too? Is it significant?

tmccombs replied:
It's possible to do. I'm not sure if it is still relevant. The examples of caching I found for github actions didn't mention anything about this.

SrodriguezO commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,78 @@
+        with: &cache_options
+          path: |
+            ~/.ivy2/cache
+            ~/.sbt
```
do we want the full `~/.sbt` or just `~/.sbt/boot` ?

tmccombs replied:
IDK, I was just using what I saw in an example.

SrodriguezO replied:
sgtm; asking cause the cache we have on travis uses the latter

SrodriguezO commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,78 @@
+            ~/.ivy2/cache
+            ~/.sbt
+            ~/.coursier
+          key: ${{ runner.os }}-sbt-${{ hashFiles('**/*.sbt') }}-${{ matrix.scala_version }}
```
should the other directories be part of the hash key too?

tmccombs replied:
what other directories?

SrodriguezO replied:
Ah, I misread that. I thought you were using the hash of the files in the .sbt dir, but it looks like you're using the hash of all files with the .sbt extension. I'm not as familiar here, is that all we need for the cache key? 

Aside, re my original question: it doesn't even make sense for the cached directories to be part of the key cause we won't have their contents until after we do the lookup :upside_down_face:

SrodriguezO reviewed (commented):
Overall this looks good, but left some comments. Only big one is whether we should release on tags or just update existing releases. I like the release on tags approach cause it mirrors Travis more closely, but your approach makes sense too. No strong preference really. 

Also double checking that we're pubishing the right artifacts since it seems that changed too.

SrodriguezO reviewed (changes requested):
Remove anchors

SrodriguezO commented:
The buid's failing with:
`Switch failed: no subprojects list "2.13.4" (or compatible version) in crossScalaVersions setting.`

fwiw, we were using 2.13.1 in Travis CI

tmccombs commented:
Yeah, I'm not sure why the sbt commands we had before aren't working any more. I'm going to try taking a different approach.

tmccombs committed:
Upgrade dependencies

tmccombs committed:
Upgrade sbt

SrodriguezO commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,75 @@
+          PGP_SECRET: ${{ secrets.PGP_SECRET }}
+  release:
+    if: startsWith(github.ref, 'refs/tags')
+    runs-on: ubuntu-latest
```
is using ubuntu-latest instead of ubuntu-20.04 for just this job intentional?

tmccombs replied:
eh, sort of. I don't remember my exact thoughts at the time, but I'm less worried about an ubuntu upgrade braking this job than I am the others.

SrodriguezO reviewed (approved):
LGMT; it might be worthwhile to verify the two jobs that only run on non-PR pushes; you can trigger them pretty easily by pushing a test tag on this branch's HEAD.

tmccombs committed:
Fixes for github CI

- Set the sonatype credential environment
- checkout in the release job, because hub requires running in the git
repo
- Improve the cache usage

tmccombs committed:
Fix version for packages uploaded to github
